EKF/mag_fusion.cpp

```diff
@@ -44,408 +44,419 @@
+			     (P[21][17] + P[16][17] * SH_MAG[5] + P[18][17] * SH_MAG[2] - P[0][17] * (magE * SH_MAG[0] + magD * SH_MAG[3] - magN *
+					     (SH_MAG[8] - 2 * q1 * q2)) + P[1][17] * (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) - P[17][17] *
+			      (2 * q0 * q1 - 2 * q2 * q3)) - P[0][21] * (magE * SH_MAG[0] + magD * SH_MAG[3] - magN *
+					      (SH_MAG[8] - 2 * q1 * q2)) + P[1][21] * (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) + SH_MAG[5] *
+			     (P[21][16] + P[16][16] * SH_MAG[5] + P[18][16] * SH_MAG[2] - P[0][16] * (magE * SH_MAG[0] + magD * SH_MAG[3] - magN *
+					     (SH_MAG[8] - 2 * q1 * q2)) + P[1][16] * (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) - P[17][16] *
+			      (2 * q0 * q1 - 2 * q2 * q3)) + SH_MAG[2] * (P[21][18] + P[16][18] * SH_MAG[5] + P[18][18] * SH_MAG[2] - P[0][18] *
+					      (magE * SH_MAG[0] + magD * SH_MAG[3] - magN * (SH_MAG[8] - 2 * q1 * q2)) + P[1][18] *
+					      (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) - P[17][18] * (2 * q0 * q1 - 2 * q2 * q3)) -
+			     (magE * SH_MAG[0] + magD * SH_MAG[3] - magN * (SH_MAG[8] - 2 * q1 * q2)) * (P[21][0] + P[16][0] * SH_MAG[5] + P[18][0] *
+					     SH_MAG[2] - P[0][0] * (magE * SH_MAG[0] + magD * SH_MAG[3] - magN * (SH_MAG[8] - 2 * q1 * q2)) + P[1][0] *
+					     (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) - P[17][0] * (2 * q0 * q1 - 2 * q2 * q3)) - P[17][21] *
+			     (2 * q0 * q1 - 2 * q2 * q3) + (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) *
+			     (P[21][1] + P[16][1] * SH_MAG[5] + P[18][1] * SH_MAG[2] - P[0][1] * (magE * SH_MAG[0] + magD * SH_MAG[3] - magN *
+					     (SH_MAG[8] - 2 * q1 * q2)) + P[1][1] * (magE * SH_MAG[4] + magD * SH_MAG[7] + magN * SH_MAG[1]) - P[17][1] *
+			      (2 * q0 * q1 - 2 * q2 * q3)));
+
+	// check for a badly conditioned covariance matrix
+	if (_mag_innov_var[2] >= R_MAG) {
+		// the innovation variance contribution from the state covariances is non-negative - no fault
+		_fault_status.bad_mag_z = false;
+
+	} else {
+		// the innovation variance contribution from the state covariances is negtive which means the covariance matrix is badly conditioned
+		_fault_status.bad_mag_z = true;
+		// we need to reinitialise the covariance matrix and abort this fusion step
+		initialiseCovariance();
+		return;
+	}
+
+
+	// Perform an innovation consistency check on each measurement and if one axis fails
+	// do not fuse any data from the sensor because the most common errors affect multiple axes.
+	_mag_healthy = true;
+
+	for (uint8_t index = 0; index <= 2; index++) {
+		_mag_test_ratio[index] = sq(_mag_innov[index]) / (sq(math::max(_params.mag_innov_gate, 1.0f)) * _mag_innov_var[index]);
+
+		if (_mag_test_ratio[index] > 1.0f) {
+			_mag_healthy = false;
+		}
+	}
+
+	if (!_mag_healthy) {
+		return;
+	}
+
+	// update the states and covariance usinng sequential fusion of the magnetometer components
+	for (uint8_t index = 0; index <= 2; index++) {
```

Comment:
APM does fuse them on the same prediction step with EKF2. The original EKF spread them across prediction steps due to concerns about timing jitter, however we were able to reduce the fusion time for EKF2. What APM does do is delay the fusion of other measurements until the next frame. This also allows for effective load levelling across multiple instances by offsetting the fusion time horizon by a few frames for each additional instance.